extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/data/input/kafkainput/KafkaInputReader.java

```diff
@@ -0,0 +1,182 @@
+        // Key currently only takes the first row and ignores the rest.
+        if (keyIterator.hasNext()) {
+          MapBasedInputRow keyRow = (MapBasedInputRow) keyIterator.next();
+          mergeList.put(
+              keyLabelPrefix + DEFAULT_KEY_STRING,
+              keyRow.getEvent().entrySet().stream().findFirst().get().getValue()
+          );
+        }
+      }
+      catch (Exception e) {
+        if (e instanceof IOException) {
+          log.error(e, "Encountered IOException during key parsing.");
+          throw (IOException) e;
+        } else if (e instanceof ParseException) {
+          log.error(e, "Encountered key parsing exception.");
+        } else {
+          log.error(e, "Encountered exception during key parsing.");
+          throw e;
+        }
+      }
+    }
+
+    List<InputRow> rows = new ArrayList<>();
+
+    // Return type for the value parser should be of type MapBasedInputRow
+    // Parsers returning other types are not compatible currently.
+    if (valueParser != null) {
+      try (CloseableIterator<InputRow> iterator = valueParser.read()) {
+        while (iterator.hasNext()) {
+      /* Currently we prefer payload attributes if there is a collision in names.
+          We can change this beahvior in later changes with a config knob. This default
+          behavior lets easy porting of existing inputFormats to the new one without any changes.
+       */
+          MapBasedInputRow row = (MapBasedInputRow) iterator.next();
+          Map<String, Object> event = new HashMap<>(mergeList);
+          event.putAll(row.getEvent());
+
+          HashSet<String> newDimensions = new HashSet<String>(row.getDimensions());
+          newDimensions.addAll(mergeList.keySet());
+          // Remove the dummy timestamp added in KafkaInputFormat
+          newDimensions.remove("__kif_auto_timestamp");
+
+          final List<String> schemaDimensions = inputRowSchema.getDimensionsSpec().getDimensionNames();
+          final List<String> dimensions;
+          if (!schemaDimensions.isEmpty()) {
+            dimensions = schemaDimensions;
+          } else {
+            dimensions = Lists.newArrayList(
+                Sets.difference(newDimensions, inputRowSchema.getDimensionsSpec().getDimensionExclusions())
+            );
+          }
+          rows.add(new MapBasedInputRow(
+              inputRowSchema.getTimestampSpec().extractTimestamp(event),
+              dimensions,
+              event
+          ));
+        }
+      }
+      catch (Exception e) {
+        if (e instanceof IOException) {
+          log.error(e, "Encountered IOException during value parsing.");
+          throw (IOException) e;
+        } else if (e instanceof ParseException) {
+          log.error(e, "Encountered value parsing exception.");
+        } else {
+          log.error(e, "Encountered exception during value parsing.");
+          throw e;
+        }
+      }
```

Comment:
if there are no other checked exception than IOException, I think we can remove this entire block and let exceptions bubble up.